### PR TITLE
Add more appdomain apis

### PR DESCRIPTION
**Pull request by rahku**

### CODE REVIEW

**Comment by rahku** [comment]
PTAL @kouvel @jkotas @weshaggard @stephentoub

File: src/System.Runtime.Extensions/src/System/AppDomain.cs [hunk 1]
```diff
@@ -87,17 +94,17 @@ public static AppDomain CreateDomain(string friendlyName)
...
-            /* TODO Assembly.LoadFrom is not implemented yet
-            Assembly assembly = Assembly.LoadFrom(path, hashValue, hashAlgorithm);
+            Assembly assembly = Assembly.LoadFrom(assemblyFile);
             return ExecuteAssembly(assembly, args);
-            */
-            return 0;
+        }
+        public int ExecuteAssembly(string assemblyFile, string[] args, byte[] hashValue, Configuration.Assemblies.AssemblyHashAlgorithm hashAlgorithm)
+        {
+            throw new PlatformNotSupportedException();
```
Comment (danmosemsft) [inline src/System.Runtime.Extensions/src/System/AppDomain.cs]: why is this PNSE? It's part of NS2.0

File: src/System.Runtime.Extensions/tests/System/AppDomainTests.cs [hunk 2]
```diff
@@ -359,13 +382,35 @@ public void SetShadowCopyPath()
         [Fact]
         public void GetAssemblies()
         {
-            // TODO
+            Assembly[] assemblies = AppDomain.CurrentDomain.GetAssemblies();
+            Assert.NotNull(assemblies);
+            Assert.True(assemblies.Length > 0);
+            AppDomain.CurrentDomain.Load(assemblies[0].GetName().FullName);
+            Assembly[] assemblies1 = AppDomain.CurrentDomain.GetAssemblies();
```
Comment (danmosemsft) [inline src/System.Runtime.Extensions/tests/System/AppDomainTests.cs]: should you be asserting what exactly you get back

File: src/System.Security.Principal/src/System/Security/Principal/PrincipalPolicy.cs [hunk 3]
```diff
@@ -0,0 +1,22 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+//
```
Comment (danmosemsft) [inline src/System.Security.Principal/src/System/Security/Principal/PrincipalPolicy.cs]: extra empty lines

File: src/System.Runtime.Extensions/src/System.Runtime.Extensions.csproj [hunk 4]
```diff
@@ -384,6 +384,11 @@
     <ProjectReference Include="..\..\System.Diagnostics.Debug\src\System.Diagnostics.Debug.csproj">
       <UndefineProperties>%(ProjectReference.UndefineProperties);TargetGroup</UndefineProperties>
     </ProjectReference>
+    <ProjectReference Include="..\..\System.Security.Principal\src\System.Security.Principal.csproj">
```
Comment (kouvel) [inline src/System.Runtime.Extensions/src/System.Runtime.Extensions.csproj]: I have added this to issue #12850 to track cleaning these up after stopping to expose those types from CoreLib.

File: src/System.Runtime.Extensions/src/System.Runtime.Extensions.csproj [hunk 5]
```diff
@@ -384,6 +384,11 @@
...
+      <Aliases>System_Security_Principal</Aliases>
+    </ProjectReference>
+    <ProjectReference Include="..\..\System.Runtime\src\System.Runtime.csproj">
```
Comment (kouvel) [inline src/System.Runtime.Extensions/src/System.Runtime.Extensions.csproj]: Why is this necessary? Wouldn't there be type conflicts between this and CoreLib? If this is necessary to reference implementations in System.Runtime, do you need an alias for this too?

File: src/System.Runtime.Extensions/src/System/AppDomain.cs [hunk 6]
```diff
@@ -196,18 +203,28 @@ public static bool MonitoringIsEnabled
...
+            add { AssemblyLoadContext.TypeResolve += value; } 
+            remove { AssemblyLoadContext.TypeResolve -= value; } 
+        }
+        public event ResolveEventHandler ResourceResolve
+        { 
+            add { AssemblyLoadContext.ResourceResolve += value; } 
+            remove { AssemblyLoadContext.ResourceResolve -= value; } 
+        }
+        public void SetPrincipalPolicy(PrincipalPolicy policy) { }
+        public void SetThreadPrincipal(IPrincipal principal)
```
Comment (kouvel) [inline src/System.Runtime.Extensions/src/System/AppDomain.cs]: It looks like the thread principal can only be set once, and throws PolicyException on desktop. Do we need to replicate that error case?

File: src/System.Runtime.Extensions/tests/System/AppDomainTests.cs [hunk 7]
```diff
@@ -359,13 +382,35 @@ public void SetShadowCopyPath()
...
+                }
+            };
+
+            AppDomain.CurrentDomain.AssemblyLoad += handler;
+
+            Assembly.LoadFile(typeof(AppDomainTests).Assembly.Location);
+
+            Assert.True(AssemblyLoadFlag);
+
+            AppDomain.CurrentDomain.AssemblyLoad -= handler;
```
Comment (kouvel) [inline src/System.Runtime.Extensions/tests/System/AppDomainTests.cs]: Similarly in other similar cases below

File: src/System.Runtime.Extensions/src/System/AppDomain.cs (hunk 1 above)
Comment (jkotas) [inline src/System.Runtime.Extensions/src/System/AppDomain.cs]: It does not make sense to implement in .NET Core.

File: src/System.Security.Principal/ref/System.Security.Principal.cs [hunk 8]
```diff
@@ -27,4 +27,9 @@ public enum TokenImpersonationLevel
         Impersonation = 3,
         None = 0,
     }
+    public enum PrincipalPolicy {
```
Comment (kouvel) [inline src/System.Security.Principal/ref/System.Security.Principal.cs]: nit: brace on next line for consistency

File: src/System.Runtime.Extensions/src/System/AppDomain.cs (hunk 1 above)
Comment (jkotas) [inline src/System.Runtime.Extensions/src/System/AppDomain.cs]: And it has zero usage because of it is only meaningful for very specific partial trust/CAS scenarios - to protect against the case where somebody may be swapping binary files underneath you.

File: src/System.Security.Principal/ref/System.Security.Principal.cs [hunk 9]
```diff
@@ -27,4 +27,9 @@ public enum TokenImpersonationLevel
...
+        UnauthenticatedPrincipal = 0,
```
Comment (kouvel) [inline src/System.Security.Principal/ref/System.Security.Principal.cs]: Members are sorted by name in refs, makes it easier to diff

File: src/System.Runtime.Extensions/src/System/AppDomain.cs (hunk 1 above)
Comment (rahku) [inline src/System.Runtime.Extensions/src/System/AppDomain.cs]: Assemblies.LoadFrom overload api which takes assemblyHashAlgorithm also throws not supoorted exception because current AssemblyLoadContext implementation does not support checking the assemblyhash. Implementation of this would require non trivial implementation in VM and I believe the api is rarely used. I have discussed this with @jkotas and he agrees to punt this to later.

File: src/System.Runtime.Extensions/src/System.Runtime.Extensions.csproj (hunk 4 above)
Comment (rahku) [inline src/System.Runtime.Extensions/src/System.Runtime.Extensions.csproj]: Thanks

File: src/System.Runtime.Extensions/src/System.Runtime.Extensions.csproj (hunk 5 above)
Comment (rahku) [inline src/System.Runtime.Extensions/src/System.Runtime.Extensions.csproj]: Actually if I do not add that build starts to complain that IPrincipal argument is not CLS compliant. However on adding this reference the error goes away. I believe System.Threading.Thread also adds a csproj reference to System.Runtime

Comment (kouvel) [inline src/System.Runtime.Extensions/src/System.Runtime.Extensions.csproj]: Strange, @ericstj, any idea why this would happen, and why the S.R reference fixes it?

Comment (ericstj) [inline src/System.Runtime.Extensions/src/System.Runtime.Extensions.csproj]: Because the System.Security.Principal.dll is System.Runtime based, without a reference to System.Runtime the compiler cannot reason about the types defined in System.Security.Principal and understand them WRT types in System.Private.CoreLib.dll

Comment (kouvel) [inline src/System.Runtime.Extensions/src/System.Runtime.Extensions.csproj]: I see, makes sense

File: src/System.Runtime.Extensions/ref/project.json [hunk 10]
```diff
@@ -1,6 +1,7 @@
 {
   "dependencies": {
-    "System.Runtime": "4.4.0-beta-24629-01"
+    "System.Runtime": "4.4.0-beta-24629-01",
+    "System.Security.Principal": "4.4.0-beta-24629-01"
```
Comment (weshaggard) [inline src/System.Runtime.Extensions/ref/project.json]: I'm not sure I like this dependency. If we want this we should consider folding System.Security.Principal into S.Runtime.Extensions instead.

File: src/System.Runtime.Extensions/src/System.Runtime.Extensions.csproj (hunk 4 above)
Comment (weshaggard) [inline src/System.Runtime.Extensions/src/System.Runtime.Extensions.csproj]: Do we know if there is anything in CoreLib that depends on these types? In other words can we successfully remove them from CoreLib?

Comment (kouvel) [inline src/System.Runtime.Extensions/src/System.Runtime.Extensions.csproj]: I believe all of the uses of IPrincipal and PrincipalPolicy are now only in dead code, and can safely be removed

Comment (kouvel) [inline src/System.Runtime.Extensions/src/System.Runtime.Extensions.csproj]: Same for IIdentity and TokenImpersonationLevel

File: src/System.Runtime.Extensions/tests/System/AppDomainTests.cs [hunk 11]
```diff
@@ -359,13 +382,35 @@ public void SetShadowCopyPath()
...
 
         [Fact]
         public void AssemblyLoad()
         {
-            // TODO
+            AssemblyLoadEventHandler handler = (sender, args) =>
+            {
+                if (args.LoadedAssembly.FullName.Equals(typeof(AppDomainTests).Assembly.FullName))
+                {
+                    AssemblyLoadFlag = !AssemblyLoadFlag;
```
Comment (rahku) [inline src/System.Runtime.Extensions/tests/System/AppDomainTests.cs]: fixed

File: src/System.Runtime.Extensions/src/System/AppDomain.cs (hunk 6 above)
Comment (rahku) [inline src/System.Runtime.Extensions/src/System/AppDomain.cs]: fixed

File: src/System.Runtime.Extensions/src/System/AppDomain.cs [hunk 12]
```diff
@@ -196,18 +205,37 @@ public static bool MonitoringIsEnabled
...
+                if (_DefaultPrincipal != null)
+                    throw new PolicyException(SR.AppDomain_Policy_PrincipalTwice);
+
+                _DefaultPrincipal = principal;
+            }
+        }
+        // TODO
+        public event ResolveEventHandler AssemblyResolve { add { } remove { } }
+
+        private IPrincipal _DefaultPrincipal;
```
Comment (kouvel) [inline src/System.Runtime.Extensions/src/System/AppDomain.cs]: nit: camel casing for fields

File: src/System.Runtime.Extensions/src/System/AppDomain.cs (hunk 1 above)
Comment (danmosemsft) [inline src/System.Runtime.Extensions/src/System/AppDomain.cs]: Might be worht a comment there.

File: src/System.Runtime.Extensions/src/System.Runtime.Extensions.csproj [hunk 13]
```diff
@@ -384,6 +384,14 @@
     <ProjectReference Include="..\..\System.Diagnostics.Debug\src\System.Diagnostics.Debug.csproj">
       <UndefineProperties>%(ProjectReference.UndefineProperties);TargetGroup</UndefineProperties>
     </ProjectReference>
+    <ProjectReference Include="..\..\System.Security.Principal\src\System.Security.Principal.csproj">
```
Comment (AlexGhiondea) [inline src/System.Runtime.Extensions/src/System.Runtime.Extensions.csproj]: Why reference the csproj directly and not the pkgproj?

File: src/System.Runtime.Extensions/src/System/AppDomain.cs [hunk 14]
```diff
@@ -196,18 +205,37 @@ public static bool MonitoringIsEnabled
...
-        public Assembly[] GetAssemblies() { throw null; }
-        public event AssemblyLoadEventHandler AssemblyLoad { add { } remove { } }
-        public event ResolveEventHandler AssemblyResolve { add { } remove { } }
-        public event ResolveEventHandler TypeResolve { add { } remove { } }
-        public event ResolveEventHandler ResourceResolve { add { } remove { } }
-        //public void SetPrincipalPolicy(System.Security.Principal.PrincipalPolicy policy) { }
-        /*
-        public void SetThreadPrincipal(System.Security.Principal.IPrincipal principal)
+        public Assembly[] GetAssemblies() => AssemblyLoadContext.GetLoadedAssemblies();
+        public event AssemblyLoadEventHandler AssemblyLoad
```
Comment (AlexGhiondea) [inline src/System.Runtime.Extensions/src/System/AppDomain.cs]: Do you need to explicitly implement the events here?

File: src/System.Runtime.Extensions/src/System/AppDomain.cs [hunk 15]
```diff
@@ -196,18 +205,37 @@ public static bool MonitoringIsEnabled
...
+            add { AssemblyLoadContext.ResourceResolve += value; } 
+            remove { AssemblyLoadContext.ResourceResolve -= value; } 
+        }
+        public void SetPrincipalPolicy(PrincipalPolicy policy) { }
+        public void SetThreadPrincipal(IPrincipal principal)
         {
             if (principal == null) throw new ArgumentNullException("principal");
-            throw new PlatformNotSupportedException();
-        }*/
+            lock (this) {
```
Comment (AlexGhiondea) [inline src/System.Runtime.Extensions/src/System/AppDomain.cs]: Can we avoid locking on `this`?

File: src/System.Runtime.Extensions/src/System/AppDomain.cs.bak [hunk 16]
```diff
@@ -0,0 +1,242 @@
+// Licensed to the .NET Foundation under one or more agreements.
```
Comment (AlexGhiondea) [inline src/System.Runtime.Extensions/src/System/AppDomain.cs.bak]: What is this file?

File: src/System.Runtime.Extensions/ref/project.json (hunk 10 above)
Comment (weshaggard) [inline src/System.Runtime.Extensions/ref/project.json]: Dang. Thanks for point out that System.Threading.Thread depends on System.Security.Principal. For now given that System.Security.Principal has minimal dependencies we can keep this dependency and move forward with what you are doing. System.Security.Principal will end up being a candidate for elimination when we look at all the dependencies after all the porting work is done.

File: src/System.Runtime.Extensions/src/System/AppDomain.cs.bak (hunk 16 above)
Comment (rahku) [inline src/System.Runtime.Extensions/src/System/AppDomain.cs.bak]: yeah should be removed

File: src/System.Runtime.Extensions/src/System.Runtime.Extensions.csproj (hunk 13 above)
Comment (rahku) [inline src/System.Runtime.Extensions/src/System.Runtime.Extensions.csproj]: Not sure about that...that's the way it is done for other projects. So will keep it.

Comment (kouvel) [inline src/System.Runtime.Extensions/src/System.Runtime.Extensions.csproj]: I believe it is because S.R.P is S.R-based rather than CoreLib-based for types, so we can't have references to both CoreLib and the S.R.P package due to type conflicts in S.R. I believe referencing just the runtime library allows this project to use implementations in S.R.P without pulling in the whole set of types. At least that's my limited understanding.

File: src/System.Runtime.Extensions/src/System/AppDomain.cs (hunk 15 above)
Comment (jkotas) [inline src/System.Runtime.Extensions/src/System/AppDomain.cs]: Right. Also, you will likely get fxcop bug on it sooner or later.

**Comment by rahku** [comment]
Thanks for the review. Anymore feedback...I plan to merge once tests complete.
